### ISSUE TITLE
FasterRCNN exampple: box_loss' moving average is Nan, but box_loss is not

### ISSUE THREAD

**Issue by LandyGuo** [ppwwyyxx/tensorpack#669]
Usage Questions Only:

I  try to train the FasterRCNN on my own dataset,  after several epochs' training,  the box_loss displays as NaN,  since the printed box_loss is  its moving average value, I also printed the current epoch's bbox_loss as  instant_box_loss, found that even the instant_box_loss is not  Nan, the 
moving average of it can also be Nan.  I am quite confused how this happens, could you give me some suggestions?  thanks 

below is the log  when bbox_loss becomes Nan first time:

^[[32m[0212 23:06:11 @saver.py:84]^[[0m Model saved to train_log/8GPU_union_debug_nan/model-224000.
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m GPUUtil/0: 90.277
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m GPUUtil/1: 91.669
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m GPUUtil/2: 91.885
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m GPUUtil/3: 92.1
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m GPUUtil/4: 90.085
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m GPUUtil/5: 91.692
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m GPUUtil/6: 91.331
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m GPUUtil/7: 92.162
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m QueueInput/queue_size: 45.419
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m fastrcnn_losses/box_loss: nan
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m fastrcnn_losses/instant_box_loss: 0.030283
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m fastrcnn_losses/label_loss: 0.11974
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m fastrcnn_losses/label_metrics/accuracy: 0.96061
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m fastrcnn_losses/label_metrics/false_negative: 0.077688
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m fastrcnn_losses/label_metrics/fg_accuracy: 0.88573
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m learning_rate: 0.01
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m maskrcnn_loss/accuracy: 0.52176
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m maskrcnn_loss/fg_pixel_ratio: 0.12285
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m maskrcnn_loss/maskrcnn_loss: 2.0157
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m maskrcnn_loss/pos_accuracy: 0.1141
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/box_loss: 0.036864
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/label_loss: 0.02364
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/label_metrics/precision_th0.1: 0.94907
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/label_metrics/precision_th0.2: 0.96451
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/label_metrics/precision_th0.5: 0.98192
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/label_metrics/recall_th0.1: 0.98504
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/label_metrics/recall_th0.2: 0.9822
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/label_metrics/recall_th0.5: 0.98012
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/num_pos_anchor/Cast: 65.971
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m rpn_losses/num_valid_anchor: 256
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m sample_fast_rcnn_targets/num_bg: 204.83
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m sample_fast_rcnn_targets/num_fg: 51.165
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m sample_fast_rcnn_targets/proposal_metrics/best_iou_per_gt: 0.82887
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m sample_fast_rcnn_targets/proposal_metrics/recall_iou0.3: 1
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m sample_fast_rcnn_targets/proposal_metrics/recall_iou0.5: 0.99929
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m total_cost: nan
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m true_mask_loss/Merge: 0.037216
^[[32m[0212 23:06:11 @monitor.py:363]^[[0m wd_cost/Merge: 0.15195

**Comment by ppwwyyxx** [comment]
Have you updated the fasterrcnn implementation to latest?

**Comment by ppwwyyxx** [comment]
Oh there was a similar issue before. This doesn't affect training. What happens is at some iteration there could be no valid boxes so box loss became NaN. Then the moving average stays NaN after that.

**Comment by ppwwyyxx** [comment]
A trivial fix would be to check for empty boxes, e.g. https://github.com/ppwwyyxx/tensorpack/commit/7f55d502aa181073f6ca0b9f2d4131c2c87369c7#diff-56a46708ffaa79088f015119c942d470

**Comment by ppwwyyxx** [comment]
I didn't think about it thoroughly, but very likely you provide some training data without valid foreground boxes, causing NaN. However this was assured by the current data processing code. I think I'll just add an assertion somewhere.

**Comment by LandyGuo** [comment]
box_loss is defined at: https://github.com/ppwwyyxx/tensorpack/blob/master/examples/FasterRCNN/model.py#L429

since huber loss in tensorflow is defined as :
![image](https://user-images.githubusercontent.com/16185531/36589521-8d2d88b2-18c6-11e8-9786-9e16157ba05f.png)
it's robust and  will never cause NaN issues.
and tf.to_float(tf.shape(labels)[0]) in L431 is constant 256 since config.FASTRCNN_BATCH_PER_IM is set as 256.
so it will nerver be NaN whether with or without  valid foreground boxes, am I right ?  Is it possible the NaN issue is caused by moving average calculation?

**Comment by ppwwyyxx** [comment]
```
box_loss = tf.losses.huber_loss(
        fg_boxes, fg_box_logits, reduction=tf.losses.Reduction.SUM)
```
It will be NaN if fg_boxes is an empty tensor.

**Comment by ppwwyyxx** [comment]
I thought huber loss returns NaN on emtpy tensors. Turned out that it seems to return 0.
Another issue could be that you provide some boxes that have zero area. As a result `encode_bbox_target` will return a tensor with NaN values.

**Comment by ppwwyyxx** [comment]
Closing due to lack of activity. Feel free to reopen if you have follow-ups.

**Comment by chenyanyin** [comment]
@LandyGuo  hello, you had solved this problem, i met too. If you, Could you tell me?
